### PR TITLE
Use lazy initialization and avoid static objects

### PR DESCRIPTION
**Pull request by jowr**
Use lazy initialization and getter functions to avoid problems with the order of initialization when linking statically.

I have had some trouble with the functions that use global maps (get_param_information and friends) since they are not guaranteed to be initialised at runtime. In my case, some constants should be initialised early using these functions and I had mysterious crashes caused by empty maps. 

This PR fixes that problem with almost no overhead.

### CODE REVIEW

File: src/DataStructures.cpp [hunk 1]
```diff
@@ -153,9 +154,16 @@ class ParameterInformation
     }
 };
 
-static ParameterInformation parameter_information;
+std::unique_ptr<ParameterInformation> parameter_information_p;
+ParameterInformation& get_parameter_information() {
```
Comment (ibell) [inline src/DataStructures.cpp:158]: Can this be a const ref here?

File: src/DataStructures.cpp [hunk 2]
```diff
@@ -379,12 +389,21 @@ class PhaseInformation
         }
     }
 };
-static PhaseInformation phase_information;
+
+std::unique_ptr<PhaseInformation> phase_information_p;
+PhaseInformation& get_phase_information() {
```
Comment (ibell) [inline src/DataStructures.cpp:394]: same question as above

File: src/DataStructures.cpp [hunk 3]
```diff
@@ -437,13 +456,22 @@ class SchemeInformation {
         }
     }
 };
-static SchemeInformation scheme_information;
+
+std::unique_ptr<SchemeInformation> scheme_information_p;
+SchemeInformation& get_scheme_information() {
```
Comment (ibell) [inline src/DataStructures.cpp:461]: again

File: src/DataStructures.cpp [hunk 4]
```diff
@@ -541,9 +569,16 @@ class InputPairInformation
     }
 };
 
-static InputPairInformation input_pair_information;
+std::unique_ptr<InputPairInformation> input_pair_information_p;
+InputPairInformation& get_input_pair_information() {
```
Comment (ibell) [inline src/DataStructures.cpp:573]: etc.

File: src/DataStructures.cpp [hunk 5]
```diff
@@ -763,10 +800,17 @@ class BackendInformation
     }
 };
 
-static BackendInformation backend_information;
+std::unique_ptr<BackendInformation> backend_information_p;
+BackendInformation& get_backend_information() {
```
Comment (ibell) [inline src/DataStructures.cpp:804]: last one

**Comment by ibell** [comment]
I see the point; a similar approach was used to manage auto-registration of backends.  Once we deal with the question about const ref, I think we can merge.

**Comment by ibell** [comment]
The performance hit should only be a single check of the pointer, right?

**Comment by jowr** [comment]
Thanks for reviewing the changes!

**Pull request merged**